libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -2558,13 +2566,9 @@ void IRGeneratorForStatements::appendExternalFunctionCall(
 	templ("funSel", IRVariable(_functionCall.expression()).part("functionSelector").name());
 	templ("address", IRVariable(_functionCall.expression()).part("address").name());
 
-	// Always use the actual return length, and not our calculated expected length, if returndatacopy is supported.
-	// This ensures it can catch badly formatted input from external calls.
-	if (m_context.evmVersion().supportsReturndata())
-		templ("returnSize", "returndatasize()");
-	else
-		templ("returnSize", to_string(returnInfo.estimatedReturnSize));
-
+	templ("returnDataSize", m_context.newYulVariable());
+	templ("estimatedReturnSize", to_string(returnInfo.estimatedReturnSize));
+	templ("supportsReturnData", m_context.evmVersion().supportsReturndata());
 	templ("reservedReturnSize", returnInfo.dynamicReturnSize ? "0" : to_string(returnInfo.estimatedReturnSize));
```

Comment:
Actually I think `reservedReturnSize` and `estimatedReturnSize` are the same. Can you add an assertion to the end of the constructor of `ReturnInfo` so that whenever dynamicReturnSize is true, estimatedReturnSize is zero and then only use `estimatedReturnSize` here?